### DISCUSSION TITLE
gitlab bot only opening up to 2 (3) PRs in total lately, bug?

### DISCUSSION THREAD

**Discussion by leaumar** [renovatebot/renovate#12342] [Request Help]
I've got 7 active repositories on gitlab, and a small one containing the renovate docker app:

```yml
include:
  - project: "renovate-bot/renovate-runner"
    file: "/templates/renovate.gitlab-ci.yml"
```

I just noticed, while having several dependencies on my projects breaking on me for what I hope are unrelated reasons, that renovate seems to stop at 3 open PRs on all 7. Since all 3 are stuck in failure until I can figure out why they fail, this means dependencies are effectively not being updated at all anymore. I have not configured `prConcurrentLimit` or otherwise set any limits in my renovate configs:
```json
{
    "extends": ["config:base", ":automergeBranch", ":automergePatch", ":automergeTypes", ":preserveSemverRanges"],
    "nvm": {
        "fileMatch": ["^\\.node-version$"]
    }
}
```

See this repo as an example: https://gitlab.com/Marnes/copy-selected-links

It has 3 failing renovations:
![image](https://user-images.githubusercontent.com/3950300/138942442-38aeae75-7bcc-48b7-b557-04fb935fd330.png)

When I run `yarn up *` on the same repo today, the diff is quite big:
![image](https://user-images.githubusercontent.com/3950300/138942586-3475af18-f665-4ac2-a731-150914322fdc.png)

The bot executions mention no errors or even warnings, and the failing PRs are kept up to date with repo changes when it runs, so it clearly works.

The git history shows a suspicious pattern of almost no dependency updates anymore since around the 21st of last month.

How come renovate stops at 3 PRs? Why is almost nothing (but not _nothing_, bizarrely) being updated since over a month? Should I report this as a bug or am I missing something?

**Comment by viceice** [comment]
Please check the debug logs of the runner, should be appended to runner pipeline

**Comment by leaumar** [comment]
[renovate-log.txt](https://github.com/renovatebot/renovate/files/7420841/renovate-log.txt)

Odd, I see mentions of a branch limit of 2?

**Reply by viceice** [reply]
I won't check your full log, please find relevant parts and paste them in a code block.

Also `config:base` has a PR hourly limit of two.

https://docs.renovatebot.com/presets-config/

**Reply by leaumar** [reply]
it's a debug log and I have no idea what's wrong, surely I can't decide which part is relevant?

And ok, that hourly limit makes sense, but that doesn't explain why it has stopped doing much of anything for almost a month now while it was super active before, including making more PRs in subsequent runs.

(the grey lines in this screenshot are weeks)
![image](https://user-images.githubusercontent.com/3950300/138946501-091fdb0d-7369-47c2-abcc-c32a29b461c1.png)

**Reply by viceice** [reply]
Check for warnings and errors.

Check the package files with updates debug message.

**Reply by leaumar** [reply]
It consistently says there are 0 errors or warnings.

I can't be sure but these debug lines seem to say it correctly recognizes the hourly (2) and total (20) limits:
```
"msg":"Processing 4 branches: renovate/cssnano-5.x, renovate/eslint-8.x, renovate/node-16.x, renovate/node-17.x","time":"2021-10-26T06:04:06.062Z","v":0}
"msg":"Calculating hourly PRs remaining","time":"2021-10-26T06:04:06.062Z","v":0}
"msg":"currentHourStart=2021-10-26T06:00:00.000+00:00","time":"2021-10-26T06:04:07.432Z","v":0}
"msg":"PR hourly limit remaining: 2","time":"2021-10-26T06:04:07.451Z","v":0}
"msg":"Calculating prConcurrentLimit (20)","time":"2021-10-26T06:04:07.451Z","v":0}
"msg":"getBranchPr(renovate/node-16.x)","time":"2021-10-26T06:04:07.451Z","v":0}
"msg":"findPr(renovate/node-16.x, undefined, open)","time":"2021-10-26T06:04:07.451Z","v":0}
"msg":"getBranchPr(renovate/node-17.x)","time":"2021-10-26T06:04:07.451Z","v":0}
"msg":"findPr(renovate/node-17.x, undefined, open)","time":"2021-10-26T06:04:07.451Z","v":0}
"msg":"getPr(271)","time":"2021-10-26T06:04:07.451Z","v":0}
"msg":"getMR(271)","time":"2021-10-26T06:04:07.452Z","v":0}
"msg":"getBranchStatus(renovate/node-17.x)","time":"2021-10-26T06:04:07.916Z","v":0}
"msg":"Got res with 1 results","time":"2021-10-26T06:04:08.816Z","v":0}
"msg":"getBranchPr(renovate/cssnano-5.x)","time":"2021-10-26T06:04:08.816Z","v":0}
"msg":"findPr(renovate/cssnano-5.x, undefined, open)","time":"2021-10-26T06:04:08.817Z","v":0}
"msg":"getBranchPr(renovate/eslint-8.x)","time":"2021-10-26T06:04:08.817Z","v":0}
"msg":"findPr(renovate/eslint-8.x, undefined, open)","time":"2021-10-26T06:04:08.817Z","v":0}
"msg":"getPr(267)","time":"2021-10-26T06:04:08.817Z","v":0}
"msg":"getMR(267)","time":"2021-10-26T06:04:08.817Z","v":0}
"msg":"getBranchStatus(renovate/eslint-8.x)","time":"2021-10-26T06:04:09.048Z","v":0}
"msg":"Got res with 1 results","time":"2021-10-26T06:04:09.270Z","v":0}
"msg":"2 PRs are currently open","time":"2021-10-26T06:04:09.270Z","v":0}
"msg":"PR concurrent limit remaining: 18","time":"2021-10-26T06:04:09.270Z","v":0}
"prsRemaining":2,"msg":"Calculated maximum PRs remaining this run","time":"2021-10-26T06:04:09.270Z","v":0}
"msg":"PullRequests limit = 2","time":"2021-10-26T06:04:09.271Z","v":0}
"msg":"Calculating hourly PRs remaining","time":"2021-10-26T06:04:09.271Z","v":0}
"msg":"currentHourStart=2021-10-26T06:00:00.000+00:00","time":"2021-10-26T06:04:09.271Z","v":0}
"msg":"PR hourly limit remaining: 2","time":"2021-10-26T06:04:09.283Z","v":0}
"msg":"Calculating branchConcurrentLimit (20)","time":"2021-10-26T06:04:09.283Z","v":0}
"msg":"2 already existing branches found: renovate/node-17.x,renovate/eslint-8.x","time":"2021-10-26T06:04:09.283Z","v":0}
"msg":"Branch concurrent limit remaining: 18","time":"2021-10-26T06:04:09.283Z","v":0}
"branchesRemaining":2,"msg":"Calculated maximum branches remaining this run","time":"2021-10-26T06:04:09.283Z","v":0}
"msg":"Branches limit = 2","time":"2021-10-26T06:04:09.283Z","v":0}
```

So question remains why it won't open any new PRs in later runs. I can't find any issues on my end tbh, and its behavior just changed pretty abruptly one day, so I really feel it's a bug.

**Reply by rarkins** [reply]
Identify one particular branch you're sure should have an MR created and then filter the logs for that branch.

**Reply by leaumar** [reply]
It's not making branches but stopping at PRs, it's not even making the branches or listing the updateable dependencies:

> "msg":"5 flattened updates found: node, node, cssnano, eslint, node","time":"2021-10-26T06:03:58.475Z","v":0}

There should be more than those 3

**Reply by leaumar** [reply]
Hang on, I did unpin my dependency versions just when this behavior began, for reasons outside renovation:  
![image](https://user-images.githubusercontent.com/3950300/139310004-7a2ad06f-640c-4776-9745-f1dab101df74.png)

So that actually shifted dependency updates _within_ the ranges to lockfile maintenance instead. Probably explains why that one's been failing so hard all of a sudden.

Right, sorry to waste your time, user forgetfulness after all. I'll be searching for either a config option to update semver ranges even within the existing range, or re-pinning the versions. Thanks guys,

**Reply by viceice** [reply]
https://docs.renovatebot.com/configuration-options/#rangestrategy 😉

**Reply by leaumar** [reply]
Yeah, was just reading that, thanks. I recall I was reading that a month ago to see if/how I could preserve renovate's behavior (an automerge/pr for every bump) but have yarn's standard semver ranges in my package.json.

Reading it now, I believe `bump` is what I need here. Going on that assumption (I'll update my repos later), I'd like to suggest the doc could perhaps be clearer. It says

> replace = Replace the range with a newer one if the new version falls outside it, e.g. ^1.0.0 -> ^2.0.0

I read this as "it will leave the semver range in package.json alone, *but still bump the dependency within that range via a lockfile update*. If the latest is out of range, it will replace the range to match the bump." The emphasized part being what one naturally assumes Renovate to do whenever it can. Modifying the semver range is a different thing from modifying the resolution of it.

Perhaps reword like so?

> replace = Replace the range with a newer one if the new version falls outside it, and update nothing otherwise.

Come to think of it now, the idea I had of `replace` seems to be what `update-lockfile` is. These definitions rely on comparing them to the other definitions to understand them correctly. I'm not sure that's wise in a list of 6, it's quite a dense bunch. :/
(inb4 "*you're* dense" ^^)

**Reply by HonkingGoose** [reply]
See PR #12384 for the wording change. 😉
